### PR TITLE
Add JSM (ESM) support for next/lib/find-config

### LINKED ISSUE #34448: `next/lib/find-config` does not support ESM config files
### Verify canary release

- [X] I verified that the issue exists in Next.js canary release

### Provide environment information

```
Operating System:
      Platform: linux
      Arch: x64
      Version: #110-Ubuntu SMP PREEMPT Thu Jan 13 19:01:34 UTC 2022
    Binaries:
      Node: 16.10.0
      npm: 7.24.0
      Yarn: 1.22.17
      pnpm: N/A
    Relevant packages:
      next: 12.0.11-canary.19
      react: 17.0.2
      react-dom: 17.0.2
```

### What browser are you using? (if relevant)

_No response_

### How are you deploying your application? (if relevant)

_No response_

### Describe the Bug

Converting Next project to `"type": "module"`, uses Tailwind and PostCSS.  `next/lib/find-config` does not attempt to load via dynamic import and so throws even after converting PostCSS config to ESM:

```
./node_modules/next/dist/build/webpack/loaders/css-loader/src/index.js??ruleSet[1].rules[2].oneOf[8].use[1]!./node_modules/next/dist/build/webpack/loaders/postcss-loader/src/index.js??ruleSet[1].rules[2].oneOf[8].use[2]!./src/styles/index.css
Error [ERR_REQUIRE_ESM]: require() of ES Module /path/to/my/repo/components/postcss.config.js from /path/to/my/repo/components/node_modules/next/dist/lib/find-config.js not supported.
Instead change the require of postcss.config.js in /path/to/my/repo/components/node_modules/next/dist/lib/find-config.js to a dynamic import() which is available in all CommonJS modules.
```

### Expected Behavior

Should attempt to load configs via ESM-CJS-interoperable default import. 

### To Reproduce

Convert a Tailwind Next.js project to `"type": "module"` and run `next dev`.

### PR DESCRIPTION
**Pull request by phanect**
Fixes #34448

Before this PR, next/lib/find-config fails to load \*.config.mjs files and \*.config.js files when `"type": "module"` is set in package.json. It expects CommonJS files although the \*.config.{js|mjs} files are written in JS modules format (i.e. using `import` and `export`).
This PR fixes it so that it can load configs written in JS modules format.

### CODE REVIEW

**Comment by phanect** [comment]
Rebased.

**Comment by ijjk** [comment]
Allow CI Workflow Run

- [ ] approve CI run for commit: 4398a00e931abb975f15103522dcfeb2101e98f3

Note: this should only be enabled once the PR is ready to go and can only be enabled by a maintainer

**Comment by ijjk** [comment]
## Failing test suites 
Commit: 4398a00e931abb975f15103522dcfeb2101e98f3

`pnpm test-dev packages/next/src/lib/find-config.test.ts`
- findConfig() > should not append message several times
- findConfig() > should not append message several times
- findConfig() > should not append message several times
- findConfig() > should not append message several times
- findConfig() > should not append message several times

<details>
<summary>Expand output</summary>

  ● findConfig() › should not append message several times

    expect(received).toStrictEqual(expected) // deep equality

    - Expected  - 0
    + Received  + 2

      Object {
    +   "default": Object {
          "basePath": "/docs",
    +   },
      }

      77 |         // Test
      78 |         const actualConfig = await findConfig(tmpDir, 'awsome')
    > 79 |         expect(actualConfig).toStrictEqual(exampleConfig)
         |                              ^
      80 |       })
      81 |     }
      82 |   }

      at Object.toStrictEqual (../packages/next/src/lib/find-config.test.ts:79:30)

  ● findConfig() › should not append message several times

    expect(received).toStrictEqual(expected) // deep equality

    - Expected  - 0
    + Received  + 3

      Object {
        "basePath": "/docs",
    +   "default": Object {
    +     "basePath": "/docs",
    +   },
      }

      77 |         // Test
      78 |         const actualConfig = await findConfig(tmpDir, 'awsome')
    > 79 |         expect(actualConfig).toStrictEqual(exampleConfig)
         |                              ^
      80 |       })
      81 |     }
      82 |   }

      at Object.toStrictEqual (../packages/next/src/lib/find-config.test.ts:79:30)

  ● findConfig() › should not append message several times

    expect(received).toStrictEqual(expected) // deep equality

    Expected: {"basePath": "/docs"}
    Received: null

      77 |         // Test
      78 |         const actualConfig = await findConfig(tmpDir, 'awsome')
    > 79 |         expect(actualConfig).toStrictEqual(exampleConfig)
         |                              ^
      80 |       })
      81 |     }
      82 |   }

      at Object.toStrictEqual (../packages/next/src/lib/find-config.test.ts:79:30)

  ● findConfig() › should not append message several times

    expect(received).toStrictEqual(expected) // deep equality

    - Expected  - 0
    + Received  + 2

      Object {
    +   "default": Object {
          "basePath": "/docs",
    +   },
      }

      77 |         // Test
      78 |         const actualConfig = await findConfig(tmpDir, 'awsome')
    > 79 |         expect(actualConfig).toStrictEqual(exampleConfig)
         |                              ^
      80 |       })
      81 |     }
      82 |   }

      at Object.toStrictEqual (../packages/next/src/lib/find-config.test.ts:79:30)

  ● findConfig() › should not append message several times

    expect(received).toStrictEqual(expected) // deep equality

    Expected: {"basePath": "/docs"}
    Received: null

      77 |         // Test
      78 |         const actualConfig = await findConfig(tmpDir, 'awsome')
    > 79 |         expect(actualConfig).toStrictEqual(exampleConfig)
         |                              ^
      80 |       })
      81 |     }
      82 |   }

      at Object.toStrictEqual (../packages/next/src/lib/find-config.test.ts:79:30)

</details>


Read more about building and testing Next.js in [contributing.md](https://github.com/vercel/next.js/blob/canary/contributing.md).

**Comment by ijjk** [comment]
# Stats from current PR

<details open>
<summary><strong>Default Build</strong></summary>

<br/>

<details>
<summary><strong>General</strong> Overall increase ⚠️</summary>

|  | vercel/next.js canary  | phanect/next.js fix-find-config-esm | Change |
| - | - | - | - |
| buildDuration | 13.7s | 13.6s | N/A |
| buildDurationCached | 7.4s | 6.1s | N/A |
| nodeModulesSize | 198 MB | 198 MB | ⚠️ +4.09 kB |
| nextStartRea..uration (ms) | 431ms | 441ms | N/A |

</details>

<details>
<summary><strong>Client Bundles (main, webpack)</strong></summary>

|  | vercel/next.js canary  | phanect/next.js fix-find-config-esm | Change |
| - | - | - | - |
| 2453-HASH.js gzip | 30.5 kB | 30.5 kB | N/A |
| 3304.HASH.js gzip | 181 B | 181 B | ✓ |
| 3f784ff6-HASH.js gzip | 53.7 kB | 53.7 kB | N/A |
| 8299-HASH.js gzip | 5.04 kB | 5.04 kB | N/A |
| framework-HASH.js gzip | 45.2 kB | 45.2 kB | ✓ |
| main-app-HASH.js gzip | 242 B | 242 B | ✓ |
| main-HASH.js gzip | 32.2 kB | 32.2 kB | N/A |
| webpack-HASH.js gzip | 1.68 kB | 1.68 kB | N/A |
| Overall change | 45.6 kB | 45.6 kB | ✓ |

</details>

<details>
<summary><strong>Legacy Client Bundles (polyfills)</strong></summary>

|  | vercel/next.js canary  | phanect/next.js fix-find-config-esm | Change |
| - | - | - | - |
| polyfills-HASH.js gzip | 31 kB | 31 kB | ✓ |
| Overall change | 31 kB | 31 kB | ✓ |

</details>

<details>
<summary><strong>Client Pages</strong></summary>

|  | vercel/next.js canary  | phanect/next.js fix-find-config-esm | Change |
| - | - | - | - |
| _app-HASH.js gzip | 196 B | 197 B | N/A |
| _error-HASH.js gzip | 184 B | 184 B | ✓ |
| amp-HASH.js gzip | 505 B | 505 B | ✓ |
| css-HASH.js gzip | 324 B | 325 B | N/A |
| dynamic-HASH.js gzip | 2.5 kB | 2.5 kB | N/A |
| edge-ssr-HASH.js gzip | 258 B | 258 B | ✓ |
| head-HASH.js gzip | 352 B | 352 B | ✓ |
| hooks-HASH.js gzip | 370 B | 371 B | N/A |
| image-HASH.js gzip | 4.21 kB | 4.21 kB | ✓ |
| index-HASH.js gzip | 259 B | 259 B | ✓ |
| link-HASH.js gzip | 2.67 kB | 2.67 kB | N/A |
| routerDirect..HASH.js gzip | 314 B | 312 B | N/A |
| script-HASH.js gzip | 386 B | 386 B | ✓ |
| withRouter-HASH.js gzip | 309 B | 309 B | ✓ |
| 1afbb74e6ecf..834.css gzip | 106 B | 106 B | ✓ |
| Overall change | 6.57 kB | 6.57 kB | ✓ |

</details>

<details>
<summary><strong>Client Build Manifests</strong></summary>

|  | vercel/next.js canary  | phanect/next.js fix-find-config-esm | Change |
| - | - | - | - |
| _buildManifest.js gzip | 481 B | 484 B | N/A |
| Overall change | 0 B | 0 B | ✓ |

</details>

<details>
<summary><strong>Rendered Page Sizes</strong></summary>

|  | vercel/next.js canary  | phanect/next.js fix-find-config-esm | Change |
| - | - | - | - |
| index.html gzip | 528 B | 529 B | N/A |
| link.html gzip | 540 B | 542 B | N/A |
| withRouter.html gzip | 523 B | 523 B | ✓ |
| Overall change | 523 B | 523 B | ✓ |

</details>

<details>
<summary><strong>Edge SSR bundle Size</strong></summary>

|  | vercel/next.js canary  | phanect/next.js fix-find-config-esm | Change |
| - | - | - | - |
| edge-ssr.js gzip | 95.1 kB | 95.1 kB | N/A |
| page.js gzip | 3.05 kB | 3.05 kB | N/A |
| Overall change | 0 B | 0 B | ✓ |

</details>

<details>
<summary><strong>Middleware size</strong></summary>

|  | vercel/next.js canary  | phanect/next.js fix-find-config-esm | Change |
| - | - | - | - |
| middleware-b..fest.js gzip | 626 B | 624 B | N/A |
| middleware-r..fest.js gzip | 151 B | 151 B | ✓ |
| middleware.js gzip | 25.5 kB | 25.5 kB | N/A |
| edge-runtime..pack.js gzip | 839 B | 839 B | ✓ |
| Overall change | 990 B | 990 B | ✓ |

</details>

<details>
<summary><strong>Next Runtimes</strong></summary>

|  | vercel/next.js canary  | phanect/next.js fix-find-config-esm | Change |
| - | - | - | - |
| app-page-exp...dev.js gzip | 171 kB | 171 kB | ✓ |
| app-page-exp..prod.js gzip | 96.6 kB | 96.6 kB | ✓ |
| app-page-tur..prod.js gzip | 98.3 kB | 98.3 kB | ✓ |
| app-page-tur..prod.js gzip | 92.8 kB | 92.8 kB | ✓ |
| app-page.run...dev.js gzip | 149 kB | 149 kB | ✓ |
| app-page.run..prod.js gzip | 91.3 kB | 91.3 kB | ✓ |
| app-route-ex...dev.js gzip | 21.3 kB | 21.3 kB | ✓ |
| app-route-ex..prod.js gzip | 15 kB | 15 kB | ✓ |
| app-route-tu..prod.js gzip | 15 kB | 15 kB | ✓ |
| app-route-tu..prod.js gzip | 14.8 kB | 14.8 kB | ✓ |
| app-route.ru...dev.js gzip | 21 kB | 21 kB | ✓ |
| app-route.ru..prod.js gzip | 14.8 kB | 14.8 kB | ✓ |
| pages-api-tu..prod.js gzip | 9.52 kB | 9.52 kB | ✓ |
| pages-api.ru...dev.js gzip | 9.8 kB | 9.8 kB | ✓ |
| pages-api.ru..prod.js gzip | 9.52 kB | 9.52 kB | ✓ |
| pages-turbo...prod.js gzip | 22.3 kB | 22.3 kB | ✓ |
| pages.runtim...dev.js gzip | 22.9 kB | 22.9 kB | ✓ |
| pages.runtim..prod.js gzip | 22.3 kB | 22.3 kB | ✓ |
| server.runti..prod.js gzip | 50.6 kB | 50.6 kB | ✓ |
| Overall change | 948 kB | 948 kB | ✓ |

</details>

<details>
<summary><strong>build cache</strong></summary>

|  | vercel/next.js canary  | phanect/next.js fix-find-config-esm | Change |
| - | - | - | - |
| 0.pack gzip | 1.56 MB | 1.56 MB | N/A |
| index.pack gzip | 105 kB | 106 kB | N/A |
| Overall change | 0 B | 0 B | ✓ |

</details>

<details>
<summary><strong>Diff details</strong></summary>

<details>
<summary>Diff for <strong>middleware.js</strong></summary>

Diff too large to display
</details>

</details>

</details>
Commit: 5243ce4c7db52d64ee1bc1b15d4c5746c344c410

File: packages/next/src/lib/find-config.ts
Comment (huozhi) [inline packages/next/src/lib/find-config.ts:40]:
```suggestion
  let isESM = false
```

File: packages/next/src/lib/find-config.ts
Comment (huozhi) [inline packages/next/src/lib/find-config.ts:54]:
```suggestion
        isESM = true
```

File: packages/next/src/lib/find-config.ts
Comment (huozhi) [inline packages/next/src/lib/find-config.ts:60]:
```suggestion
    } catch {
```

File: packages/next/src/lib/find-config.ts
Comment (huozhi) [inline packages/next/src/lib/find-config.ts:69]:
```suggestion
      return isESM ? (await import(filePath)).default : require(filePath)
```

File: packages/next/src/lib/find-config.test.ts
Comment (huozhi) [inline packages/next/src/lib/find-config.test.ts:33]: Maybe separate `package.json` into a new test suite instead mixing with extensions? Will look more clear

**Comment by phanect** [comment]
@huozhi Thanks for the review! I will update the code.

**Comment by phanect** [comment]
@huozhi I've applied your suggestions.
I have rebased this branch by `canary` branch, but I didn't touch the commits before f23ce38 (test: support to load config when "type": "module" is set.

File: packages/next/src/lib/find-config.test.ts
Comment (phanect) [reply packages/next/src/lib/find-config.test.ts]: Thanks, I've applied in [17f4f6f](https://github.com/vercel/next.js/pull/63109/commits/17f4f6f338b5e50639818bcb7ad83e8a7bfa418a). Can you check if my change suits your expectation?

**Review by huozhi — APPROVED** [review]
Awesome! Thank you!

**Comment by phanect** [comment]
@huozhi I noticed GHA job failed: https://github.com/vercel/next.js/actions/runs/8305434996/job/22735976306?pr=63109
The CI seems to fail on downloading https://cdn.ampproject.org/v0/validator_wasm.js and it might be just a network error.

I rebased this branch and force-pushed to re-run the CI, but I forgot that it requires your approval to run GHA. Can you approve it?

**Comment by huozhi** [comment]
Some might be flaky, please don't re-push again since we need to re-approve and re-run the CI. I'll monitor the CI and land this once it's passed 🙏

**Comment by phanect** [comment]
Thanks for merging!

**Comment by phanect** [comment]
@huozhi Thanks for merging!

> no need to re-push again since we need to re-approve and re-run the CI.

Sorry for that 🙏
I will refrain from pushing again when not necessary from next time.

**Comment by karlhorky** [comment]
Thanks for the PR for ESM support @phanect and for the review and merge @huozhi 🙌 

I can confirm that [`next@14.2.0-canary.25`](https://github.com/vercel/next.js/releases/tag/v14.2.0-canary.25) does have support for the following ESM configurations 🎉

1. CommonJS project + `postcss.config.mjs` in ESM format
2. ESM project + `postcss.config.js` in ESM format

More details here:

- https://github.com/vercel/next.js/issues/34448#issuecomment-2002513315

**Pull request merged**
